src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -58,35 +86,123 @@ internal sealed class ExternalErrorDiagnosticUpdateSource : IDiagnosticUpdateSou
+        /// This API is only intended to be invoked from <see cref="ProjectExternalErrorReporter"/> while a build is in progress.
+        /// </summary>
         public bool IsSupportedDiagnosticId(ProjectId projectId, string id)
             => BuildInprogressState?.IsSupportedDiagnosticId(projectId, id) ?? false;
 
+        private void OnBuildProgressChanged(InProgressState? state, BuildProgress buildProgress)
+        {
+            if (state != null)
+            {
+                _lastBuiltResult = state.GetBuildErrors();
+            }
+
+            RaiseBuildProgressChanged(buildProgress);
+        }
+
         public void ClearErrors(ProjectId projectId)
         {
-            // capture state if it exists
+            // Capture state if it exists
             var state = BuildInprogressState;
 
-            _taskQueue.ScheduleTask(nameof(ClearErrors), () =>
+            // Update the state to clear diagnostics and raise corresponding diagnostic updated events
+            // on a serialized task queue.
+            _taskQueue.ScheduleTask(nameof(ClearErrors), async () =>
             {
-                // this will get called if the project is actually built by "build" command.
-                // we track what project has been built, so that later we can clear any stale live errors
-                // if build give us no errors for this project
-                state?.Built(projectId);
+                if (state == null)
+                {
+                    // TODO: Is it possible that ClearErrors can be invoked while the build is not in progress?
+                    // We fallback to current solution in the workspace and clear errors for the project.
+                    await ClearErrorsCoreAsync(projectId, _workspace.CurrentSolution, state).ConfigureAwait(false);
+                }
+                else
+                {
+                    // We are going to clear the diagnostics for the current project.
+                    // Additionally, we clear errors for all projects that transitively depend on this project.
+                    // Otherwise, fixing errors in core projects in dependency chain will leave back stale diagnostics in dependent projects.
 
-                ClearProjectErrors(state?.Solution ?? _workspace.CurrentSolution, projectId);
+                    // First check if we already cleared the diagnostics for this project when processing a referenced project.
+                    // If so, we don't need to clear diagnostics for it again.
+                    if (state.WereProjectErrorsCleared(projectId))
+                    {
+                        return;
+                    }
+
+                    var solution = state.Solution;
+
+                    await ClearErrorsCoreAsync(projectId, solution, state).ConfigureAwait(false);
+
+                    var transitiveProjectIds = solution.GetProjectDependencyGraph().GetProjectsThatTransitivelyDependOnThisProject(projectId);
```

Comment:
instead of clearing transitively, would it suffice to clear only direct dependencies?  then clearing those would clear their dependencies, and so on.  so you would get transitivity for free?

Reply:
That will not handle cases like following:

1. Assume 3 projects A, B and C, such that A references B and B references C, so A transitively depends on C.
2. Introduce an API level syntax error in C
3. Build solution - you will get errors in A, B and C, some of which are build-only errors.
4. Fix the error in C and just build C
5. Error list would have stale errors in A, and even opening files with these errors will not remove the error list entries for A. Intellisense will correctly resolve no errors in A, so you will not get squiggles, but build-only errors in A.